File: CSES/Graphs/Labyrinth.py
########################################################################################################################
# -----------------------------------------------AUTHOR: shank_555-----------------------------------------------------#
########################################################################################################################
import time
import bisect
import functools
import math
import os
import random
import re
import sys
import threading
from collections import Counter, defaultdict, deque
from copy import deepcopy
from functools import cmp_to_key, lru_cache, reduce
from heapq import heapify, heappop, heappush, heappushpop, nlargest, nsmallest
from io import BytesIO, IOBase
from itertools import accumulate, combinations, permutations
from operator import add, iand, ior, itemgetter, mul, xor
from string import ascii_lowercase, ascii_uppercase
from typing import *
import collections
import heapq
import itertools
alphabets = list("abcdefghijklmnopqrstuvwxyz")
vowels = list("aeiou")
MOD1 = int(1e9 + 7)
MOD2 = 998244353
INF = int(1e17)
I = lambda: input()
II = lambda: int(input())
MII = lambda: map(int, input().split())
LI = lambda: list(input().split())
LII = lambda: list(map(int, input().split()))
LGMII = lambda: map(lambda x: int(x) - 1, input().split())
LGLII = lambda: list(map(lambda x: int(x) - 1, input().split()))
inf = float('inf')
dirs = [[0, 1], [0, -1], [1, 0], [-1, 0]]  

def solve():
    n, m = map(int, input().split()) 
    grid = [list(input()) for _ in range(n)]  

    vis = [[False] * m for _ in range(n)] 
    parent = {} 
    d = {(0, 1): "R", (0, -1): "L", (1, 0): "D", (-1, 0): "U"}  

    def bfs(si, sj):
        queue = deque([(si, sj)]) 
        vis[si][sj] = True
        while queue:
            i, j = queue.popleft()
            if grid[i][j] == "B": 
                return i, j
            for di, dj in dirs:
                ni, nj = i + di, j + dj
                if 0 <= ni < n and 0 <= nj < m and not vis[ni][nj] and grid[ni][nj] != "#":
                    vis[ni][nj] = True
                    queue.append((ni, nj))
                    parent[(ni, nj)] = (i, j)  
        return -1, -1  

    si, sj, bi, bj = -1, -1, -1, -1
    for i in range(n):
        for j in range(m):
            if grid[i][j] == "A":
                si, sj = i, j
            elif grid[i][j] == "B":
                bi, bj = i, j

    ei, ej = bfs(si, sj) 
    if (ei, ej) == (-1, -1):  
        print("NO")
        return

    path = []
    cur = (ei, ej)
    while cur != (si, sj):
        prev = parent[cur]
        di, dj = cur[0] - prev[0], cur[1] - prev[1]
        path.append(d[(di, dj)])
        cur = prev

    path.reverse()  
    print("YES")
    print(len(path))
    print("".join(path))

solve()

File: CSES/Graphs/Labyrinth.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

const int MOD1 = 1e9 + 7;
const int MOD2 = 998244353;
const long long INF = 1e17;
const int dirs[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

// BFS function to find the shortest path to 'B'
pair<int, int> bfs(int si, int sj, int n, int m, const vector<vector<char>>& grid, vector<vector<bool>>& vis, map<pair<int, int>, pair<int, int>>& parent) {
    queue<pair<int, int>> q;
    q.push(make_pair(si, sj));
    vis[si][sj] = true;

    while (!q.empty()) {
        pair<int, int> current = q.front();
        q.pop();

        int i = current.first;
        int j = current.second;

        if (grid[i][j] == 'B') {
            return make_pair(i, j);
        }

        for (int k = 0; k < 4; ++k) {
            int ni = i + dirs[k][0];
            int nj = j + dirs[k][1];

            if (ni >= 0 && ni < n && nj >= 0 && nj < m && !vis[ni][nj] && grid[ni][nj] != '#') {
                vis[ni][nj] = true;
                q.push(make_pair(ni, nj));
                parent[make_pair(ni, nj)] = make_pair(i, j);
            }
        }
    }
    return make_pair(-1, -1); // No path to 'B' found
}

void solve() {
    int n, m;
    cin >> n >> m;

    vector<vector<char>> grid(n, vector<char>(m));
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            cin >> grid[i][j];
        }
    }

    vector<vector<bool>> vis(n, vector<bool>(m, false));
    map<pair<int, int>, pair<int, int>> parent;
    map<pair<int, int>, string> d;
    d[make_pair(0, 1)] = "R";
    d[make_pair(0, -1)] = "L";
    d[make_pair(1, 0)] = "D";
    d[make_pair(-1, 0)] = "U";

    int si = -1, sj = -1, bi = -1, bj = -1;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            if (grid[i][j] == 'A') {
                si = i;
                sj = j;
            } else if (grid[i][j] == 'B') {
                bi = i;
                bj = j;
            }
        }
    }

    pair<int, int> endpoint = bfs(si, sj, n, m, grid, vis, parent);
    int ei = endpoint.first, ej = endpoint.second;

    if (ei == -1 && ej == -1) {
        cout << "NO" << endl;
        return;
    }

    // Reconstruct the path from 'B' to 'A'
    vector<string> path;
    pair<int, int> cur = make_pair(ei, ej);
    while (cur != make_pair(si, sj)) {
        pair<int, int> prev = parent[cur];
        int di = cur.first - prev.first;
        int dj = cur.second - prev.second;
        path.push_back(d[make_pair(di, dj)]);
        cur = prev;
    }

    reverse(path.begin(), path.end());
    cout << "YES" << endl;
    cout << path.size() << endl;
    for (int i = 0; i < path.size(); ++i) {
        cout << path[i];
    }
    cout << endl;
}

int main() {
    solve();
    return 0;
}
